nixd/lib/Controller/Completion.cpp

```diff
@@ -298,7 +338,23 @@ void Controller::onCompletion(const CompletionParams &Params,
               completeAttrName(Scope, Prefix, Options,
                                ClientCaps.CompletionSnippets, List.items);
             }
-          } else {
+            FinishedCompletion = true;
+          }
+
+          // if we are in a literal path, use PathCompletionProvider only
+          if (!FinishedCompletion) {
+            const auto *Parent = PM.upExpr(*Desc);
+            if (Parent->kind() == Node::NK_ExprPath) {
+              const auto &Path = static_cast<const nixf::ExprPath &>(*Parent);
+              if (Path.parts().isLiteral()) {
+                completeExprPath(File, Path, List.items);
+                FinishedCompletion = true;
+              }
+            }
+          }
+
+          // otherwise, fallback to VLACompletionProvider
```

Comment:
You can determine whether or not providing VLA completion based on the `upExpr` result. For example if an `ExprPath` is encountered, then provide path completion, otherwise if `ExprVar`, provide VLA completion. Try to determine expr kind first, may make the source code more concise.

Reply:
When writing code, I believe that in future there will be more and more different completion modes added, and ultimately form something like that:

![image](https://github.com/nix-community/nixd/assets/29816865/03dfdd49-e633-4cd5-916e-022dfd24588c)

IMO the nested levels of codes will be too deep if we just adding `if`, so I refactor this part of code. But since we only have 3 completion modes at this time, I will revert changes here and use nested `if`. 

Another solution here will be some `goto`s.

Reply:
> IMO the nested levels of codes will be too deep if we just adding if, so I refactor this part of code. But since we only have 3 completion modes at this time,

These three completion modes are `switch`able by the "context" of nix codes. For example, I suppose the attrpath completion only makes sense if there is actually an attrpath user typing, and path completion only makes sense if we are in attrpath.

> Another solution here will be some `goto`s.

No, we are not writing automata.